src/main/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProvider.kt

```diff
@@ -0,0 +1,115 @@
+import org.rust.lang.core.psi.RsDotExpr
+import org.rust.lang.core.psi.RsMethodCall
+import org.rust.lang.core.psi.ext.RsElement
+import org.rust.lang.core.psi.ext.childOfType
+import org.rust.lang.core.psi.ext.endOffset
+import org.rust.lang.core.psi.ext.parentDotExpr
+import org.rust.lang.core.types.ty.Ty
+import org.rust.lang.core.types.type
+import javax.swing.JComponent
+
+class RsChainMethodTypeHintsProvider : InlayHintsProvider<RsChainMethodTypeHintsProvider.Settings> {
+    override val key: SettingsKey<Settings> get() = KEY
+
+    override val name: String get() = "Chain method hints"
+
+    override val previewText: String
+        get() = """
+            let foo: Result<Option<i32>, &str> = Ok(Some(10i32));
+            foo
+              .map_err(|_| 0_u32)
+              .unwrap()
+              .and_then(|_| Some("foo"))
+              .unwrap();
+            """.trimIndent()
+
+    override fun createConfigurable(settings: Settings): ImmediateConfigurable = object : ImmediateConfigurable {
+        override fun createComponent(listener: ChangeListener): JComponent {
+            return panel {}
+        }
+    }
+
+    override fun createSettings(): Settings = Settings()
+
+    override fun getCollectorFor(file: PsiFile, editor: Editor, settings: Settings, sink: InlayHintsSink): InlayHintsCollector? =
+        object : FactoryInlayHintsCollector(editor) {
+
+            val typeHintsFactory = RsTypeHintsPresentationFactory(factory, true)
+
+            override fun collect(element: PsiElement, editor: Editor, sink: InlayHintsSink): Boolean {
+                if (DumbService.getInstance(file.project).isDumb) return true
+                if (element !is RsElement) return true
+                if (!element.isEnabledByCfg) return true
+                if (element !is RsMethodCall) return true
+                if (!element.isLastInChain) return true
+
+                val chain = collectChain(element)
+                var lastType: Ty? = null
+                for (call in chain.dropLast(1)) {
+                    if (call.type != lastType) {
+                        presentTypeForMethodCall(call)
+                        lastType = call.type
+                    }
+                }
+
+                return true
+            }
+
+            private fun presentTypeForMethodCall(call: RsMethodCall) {
+                val project = call.project
+                val type = call.type
+                val presentation = typeHintsFactory.typeHint(type)
+                val finalPresentation = presentation.withDisableAction(project)
+                sink.addInlineElement(call.endOffset, true, finalPresentation)
+            }
+        }
+
+    private fun InlayPresentation.withDisableAction(project: Project): InsetPresentation = InsetPresentation(
+        MenuOnClickPresentation(this, project) {
+            listOf(InlayProviderDisablingAction(name, RsLanguage, project, key))
+        }, left = 1
+    )
+
+    class Settings
```

Comment:
Let's add settings similar to `Minimal unique type count to show hints` Java setting

Reply:
I'm not sure if we should copy the specific behaviour of the Java setting, IMO it behaves in a weird way, at least for Rust. For example with minimal unique types = 1 (which is the lowest value that you can set it to):
```java
a
.b()
.c();
```
This doesn't show any hints. However if `b()` returned some complex type, I think that it would still be nice to show it in Rust. In Java it's mostly for builders where the types are usually obvious, while in Rust I feel like the types are often pretty complicated and you can't even name them by yourself, so you do things like assigning them into variables of a different type just to see what the real type is from the compiler error.

Also when a type repeats it suddently stops showing types:
```java
a
 .a()
 .b()
 .b()
 .c();
```

We already ignore repeated types, so some of the noise will be reduced. Maybe we could include a setting to only hint a specific type once in a specific chain? Like this:
```rust
// include_once=false
a
.a() // hint
.b() // hint
.a() // hint
.b() // hint

// include_once=true
a
.a() // hint
.b() // hint
.a()
.b()
```

Reply:
Maybe we can also ignore obvious stuff like `.clone()`? But that would be pretty brittle I think.

Reply:
But I understood that we should have another option. Right now our implementation doesn't show type of method call if it's the same as previous type.
At the same time, we want to hide type hints for unknown types. Since our name resolution and type inference are not ideal (unfortunately), sometimes we can get a situation when it's almost impossible from the user side to distinguish "same as previous type" and "unknown type" cases.

So let's add an option to show consecutive same method types and enable it by default - it will help us to collect a lot of feedback about our type inference. After we implemented type hints, we started to get much more reports about problems in our name resolution/type inference